### ISSUE TITLE
OpenAPI 3.1.0 is coming

### ISSUE THREAD

**Issue by danielgtaylor** [getkin/kin-openapi#230]
First release candidate has been put up:

https://github.com/OAI/OpenAPI-Specification/releases/tag/3.1.0-rc0

Creating this issue to track what work needs to be done to support it. Of particular interest is the section on breaking changes to align with JSON Schema 2019-09.

**Comment by fenollp** [comment]
My opinion is it'd be best to bring in https://github.com/xeipuuv/gojsonschema and switch on OpenAPI's version field to know when to translate an OpenAPI `<v3.1` schema to the JSON Schema format that this project handles.

**Comment by kevin-postman** [comment]
I see this issue was closed and help is wanted. Now that 3.1.0 is released, are there plans to soon start adding support for 3.1.0? Is there any discussion outside of Issues that I can participate in to get involved?

**Comment by fenollp** [comment]
Hi Kevin, this issue is still open although was not discussed outside of issues as far as I'm aware. 
I think the way to support 3.1 as well as solve all the validation issues is through a `func (s *openapi3.Swagger) CompileSchemas() error` which would convert our schema data to gojsonschema's and for `(*openapi3.Schema) VisitJSON*(...) error` funcs to use that.
That'll break today's usage but this lib is `v0` anyway.

**Comment by fenollp** [comment]
Actually gojsonschema only `Supports draft-04, draft-06 and draft-07.` today.
I'm not finding any lib that supports the [2020-12 draft](https://tools.ietf.org/html/draft-bhutton-json-schema-00) that [OpenAPIv3.1](https://github.com/OAI/OpenAPI-Specification/releases/tag/3.1.0) relies on.
I do think that the move to this lib needs to happen though and support for new JSON Schema drafts will eventually land in gojsonschema AFAIK.

Do you have needs for the newer Schema Drafts or are you mostly interested in the other OpenAPIv3.1 changes?

**Comment by kevin-postman** [comment]
Me personally, interested in 3.1 support. Obviously the sooner tools support 3.1 the better and a lot of tools are built on top of kin-openapi. So I would think this is a pretty important update to add to kin-openapi as soon as possible. If moving to gojsonschema is the future anyway, then perhaps moving to it now along with adding 3.1 support should be taken up. 
My apologies on thinking it was closed.. I see that was a linked issue. :D.

**Comment by wolfeidau** [comment]
@fenollp have you had a look at https://github.com/qri-io/jsonschema as a possible alternative given it currently states it supports 2019-09 and is supposedly faster?

**Comment by fenollp** [comment]
Yes and I did see the benchmark mentioning it was less correct than the other lib. See https://github.com/wbvinc/go-jsonschema-validator-benchmarks
I did not verify this on my end however. 

I haven't dug in deeper yet but will in the coming weeks. Thanks for the link.

**Comment by timbunce** [comment]
Not only does https://github.com/qri-io/jsonschema support 2019-09, it's also [actively maintained](https://github.com/qri-io/jsonschema/pulse/monthly).

In contrast, gojsonschema had a last commit in Oct 2020 and last release in Oct 2019. Re supporting 2019-09 the author [said in March 2020](https://github.com/xeipuuv/gojsonschema/issues/289#issuecomment-598822577):
> Plans yes, but time no. The new specification also requires a thorough rewrite of all the internals in order to support things like $vocabulary and unEvaluatedProperties and the like and I haven't found the courage yet to start that project in my spare time.

The santhosh-tekuri/jsonschema repo has no commits since Feb and no sign of working towards supporting 2019-09.

I think qri-io/jsonschema is the only horse in the race at the moment. Adopting it, and contributing fixes as needed, seems like the most fruitful path.

**Comment by kevin-postman** [comment]
If moving to qri-io/jsonschema moves kin-openapi to accepting openapi 3.1.0 sooner, I am all in favor. The sooner those of us using kin-openapi can start adding support for 3.1.0 the better. Seems like one of the nagging issues since the inception of Swagger/OpenAPI is how long it takes tools to adopt/migrate to a new version. Maybe reaching out to the qri-io team and having a dialog with them regarding their commitment might ease the idea of replacing the recent work with gojsonschema?

**Comment by timbunce** [comment]
> The santhosh-tekuri/jsonschema repo has no commits since Feb and no sign of working towards supporting 2019-09.

santhosh-tekuri/jsonschema [v4.0.0](https://github.com/santhosh-tekuri/jsonschema/releases/tag/v4.0.0) says "draft 2019-09 support".

> I think qri-io/jsonschema is the only horse in the race at the moment.

Seems like it's now a two horse race.

**Comment by bojanz** [comment]
https://github.com/santhosh-tekuri/jsonschema now supports draft 2020-12, putting it ahead in the race.

**Comment by kevin-postman** [comment]
Is there any progress on 3.1 support? Wondering if anyone is working on it yet, or if not is there a plan to start work on it soon?

**Comment by peteraba** [comment]
Just a little nudge for this issue:
- https://github.com/santhosh-tekuri/jsonschema/issues/48
- https://github.com/qri-io/jsonschema/issues/113
- https://github.com/xeipuuv/gojsonschema/issues/340

**Comment by kevin-postman** [comment]
Just following up to see if there is any movement on this issue? Any work in progress by anyone? Anytime soon maybe? 3.1 has been final for almost a year.. be good to see this project get a jump on it so Go projects depending upon it can start to support 3.1.

**Comment by fenollp** [comment]
Hi @kevin-postman and others,
I understand this issue (and some others) are pressing to some and would like to apologize for the less than stellar progress here.

I have been maintaining this project for a few years now and am looking to turn this into a more sustainable initiative.
Recently, I set up GitHub Sponsorship so if you'd like to support this project's future and preserve some of my mental health I encourage you to register for a monthly sponsoring [here](https://github.com/sponsors/fenollp). A one-time donation can also help towards solving piecemeal issues.

Here's a tutorial from GitHub on [how to find other maintainers that may need your help](https://docs.github.com/en/sponsors/getting-started-with-github-sponsors/quickstart-for-finding-contributors-to-sponsor) whose work you might depend on.

Regarding this specific issue, considering the large ramifications and attention to quality it entails, I would like to work on it once some funding has been promised or secured. We can discuss terms privately.

Licensing won't change due to sponsorship. All open source work will continue to be available to all and enjoyed equally by all members of the community.

TL;DR: Nothing changes and this continues being MIT licensed. Please consider Sponsoring.

**Comment by philsturgeon** [comment]
Hello! As the maintainer of [openapi.tools](https://openapi.tools), and as somebody works with Linux Foundation helping out in OpenAPI-land, I'm reaching out to tooling vendors to track the progress towards supporting OpenAPI v3.1, to see what roadblocks there are beyond folks just generally being busy at this ridiculous time.

OpenAPI v3.1 has a bunch of great little changes, solving problems like the [JSON Schema <!=> OpenAPI Schema Object divergence](https://apisyouwonthate.com/blog/openapi-json-schema-divergence). It also fixes some other inconsistencies and duplicate ways of doing things. It's the best version and everyone should be using it, but we need tooling to catch up. Just in case folks didn't notice, or don't have resources to simplify the process, I'm here to give a friendly prod and send over some handy links.

Here are a few articles showing off the differences between OpenAPI v3.0 and v3.1.

- https://www.openapis.org/blog/2021/02/16/migrating-from-openapi-3-0-to-3-1-0
- https://www.apimatic.io/blog/2021/09/migrating-to-and-from-openapi-3-1/ 
- https://nordicapis.com/whats-new-in-openapi-3-1-0/

Here are some example files which can make for handy pass/fail test cases:

https://github.com/Mermade/openapi3-examples/tree/master/3.1 

If you're looking for the JSON Schema that defines a valid OpenAPI document, that'll be right over here: 

https://github.com/OAI/OpenAPI-Specification/tree/main/schemas/v3.1 

Hopefully these links can help your contributors step up with confidence to help you get this project updated. If you do start getting pull requests towards OpenAPI 3.1 support, please send a pull request to [openapi.tools](https://openapi.tools) with an issue tracking the progress, and folks will have a way to subscribe for updates.

LMK if you have any questions!

**Comment by santhosh-tekuri** [comment]
@fenollp you can easily compile schemas from openapi document using `github.com/santhosh-tekuri/jsonschema/v5`

here is the playground link that shows how to do it: https://go.dev/play/p/Hk077OKVi4e

**Comment by jamietanna** [comment]
Just as a quick nudge on this - this appears to be [blocking folks doing contract testing](https://gitlab.com/jamietanna/httptest-openapi/-/issues/3) using one of the libraries I maintain that depends on this.

I may start looking into this at some point to see if we can start getting - even if it's just partial - support for OpenAPI 3.1

**Comment by petkostas** [comment]
Nice to see this moving forward, the 3.1 is getting increasingly painful (at least in the GOlang world) as lots of tools rely on this library at the moment.
Let us know how we can contribute and get this moving forward.

**Comment by SVilgelm** [comment]
I have played with the generics and created a small lib for openapi v3.1 https://github.com/sv-tools/openapi

Feel free to use it as an example

**Comment by philsturgeon** [comment]
This also seems like it might be handy. github.com/pb33f/libopenapi

**Comment by jamietanna** [comment]
I've recently been chatting with @deepmap-marcinr around this with https://github.com/deepmap/oapi-codegen that we maintain.

I've looked at this a few times to start trying to pick up this, but have been a little overwhelmed by the amount of work it may lead to :sweat_smile: that may not be a fair assumption, but I've not got full appreciation of all the differences between versions.

@philsturgeon aside from the links above, do you have any reading / people that we may be able to chat to who've added support for 3.1 on top of 3.0? 

I've planned to look through things like `libopenapi` :point_up: and others to see what sort of test data they've got that may give us an indication of how much needs to change in this library to support it.

**Comment by timbunce** [comment]
[Bowtie](https://bowtie-json-schema.readthedocs.io/en/latest/) is "a meta-validator for the JSON Schema specification" with test suites for multiple versions, including draft7, 2019-09, and 2020-12.

The [results for draft7](https://bowtie-json-schema.github.io/bowtie/draft7) include [gojsonschema](https://github.com/xeipuuv/gojsonschema) with 10 errors, and [jsonschema](https://github.com/santhosh-tekuri/jsonschema) with zero errors. The [results for 2020-12](https://bowtie-json-schema.github.io/bowtie/draft2020-12) include only [jsonschema](https://github.com/santhosh-tekuri/jsonschema) with, again, zero errors.

**Comment by Ilyes512** [comment]
Is it important to keep supporting the older OpenApi implementations?

It might be easier to create a new version of this package (or a totally new one) that only starts supporting OpenApi from version 3.1, because you might be able to offload a larger part of the logic to an existing jsonschema package?

**Comment by jamietanna** [comment]
@deepmap-marcinr I think you may be talking about oapi-codegen not kin-openapi? 😁

**Comment by deepmap-marcinr** [comment]
Haha, indeed. Sorry, I got mailing lists crossed.

________________________________________

**Comment by podhmo** [comment]
https://github.com/getkin/kin-openapi/issues/230#issuecomment-1519846246

> It might be easier to create a new version of this package (or a totally new one) that only starts supporting OpenApi from version 3.1, because you might be able to offload a larger part of the logic to an existing jsonschema package?

Indeed. As a smaller step, for example, creating an openapi31 directory and reimplementing everything might be a better flow?

As for creating new package, I honestly have mixed feelings about seemingly ignoring kin-openapi's excellent contributions.

**Comment by cletter7** [comment]
Is this being worked on or abandoned?

**Comment by excavador** [comment]
Any updates?

**Comment by Malix-off** [comment]
Seems like it's time to move to https://github.com/pb33f/libopenapi

**Comment by praneetloke** [comment]
I think @fenollp is pretty clear that he would like some support from the community in terms of sponsorship to take on this task. See this [comment](https://github.com/getkin/kin-openapi/issues/230#issuecomment-1002659224) from him. It would be nice to lend him the support he is looking for especially given the fact that this library is used in so many different packages in the Go ecosystem.

**Comment by StevenACoffman** [comment]
Hey, this is a beloved and much-used library, and I appreciate all the hard work that has gone into it and I know everyone else reading this does as well.

If **you** are using it at your workplace, or are using [one of the many popular libraries and applications that depends on it](https://www.grank.dev/pkg/github.com/getkin/kin-openapi.html) then please consider doing one of the following:

1. asking **your own company** to contribute code or [sponsorship of fenollp](https://github.com/sponsors/fenollp) for this upstream library [kin-openapi](https://github.com/getkin/kin-openapi) to support OpenAPI 3.1 - see here https://github.com/getkin/kin-openapi/issues/230  and specifically [this comment](https://github.com/getkin/kin-openapi/issues/230#issuecomment-1002659224).

2. Using your company's vendor support system, report a support issue with **your company's vendor** that uses this library asking that vendor to contribute code or [sponsorship of fenollp](https://github.com/sponsors/fenollp) for this upstream library [kin-openapi](https://github.com/getkin/kin-openapi) to support OpenAPI 3.1 - see here https://github.com/getkin/kin-openapi/issues/230  and specifically [this comment](https://github.com/getkin/kin-openapi/issues/230#issuecomment-1002659224). In no particular order: Grafana, Google (Apigee), Hashicorp, PingCap, Ory, Solo, Replicate, and all the blockchain companies that own the downstream.

3. Report a GitHub issue with [one of the many company-owned popular libraries and applications that depends on this one](https://www.grank.dev/pkg/github.com/getkin/kin-openapi.html) asking **their corporate sponsor** to contribute code or [sponsorship of fenollp](https://github.com/sponsors/fenollp) for this upstream library [kin-openapi](https://github.com/getkin/kin-openapi) to support OpenAPI 3.1 - see here https://github.com/getkin/kin-openapi/issues/230  and specifically [this comment](https://github.com/getkin/kin-openapi/issues/230#issuecomment-1002659224). Do **_not_** pester other individual unpaid volunteer maintainers of other open-source projects.

> I have been maintaining this project for a few years now and am looking to turn this into a more sustainable initiative.  Recently, I set up GitHub Sponsorship so if you'd like to support this project's future and preserve some of my mental health I encourage you to register for a monthly sponsoring [here](https://github.com/sponsors/fenollp). A one-time donation can also help solve piecemeal issues.
> 
> Here's a tutorial from GitHub on [how to find other maintainers that may need your help](https://docs.github.com/en/sponsors/getting-started-with-github-sponsors/quickstart-for-finding-contributors-to-sponsor) whose work you might depend on.
> 
> Regarding this specific issue, considering the large ramifications and attention to quality it entails, I would like to work on it once some funding has been promised or secured. We can discuss terms privately.
> 
> Licensing won't change due to sponsorship. All open-source work will continue to be available to all and enjoyed equally by all community members.

Just always remember:

![dependency](https://github.com/user-attachments/assets/b215574f-6873-4ddc-b1a5-4ce0ee4989aa)

**Comment by excavador** [comment]
> Hey, this is a beloved and much-used library, and I appreciate all the hard work that has gone into it and I know everyone else reading this does as well.
> 
> If **you** are using it at your workplace, or are using [one of the many popular libraries and applications that depends on it](https://www.grank.dev/pkg/github.com/getkin/kin-openapi.html) then please consider doing one of the following:
> 
> 1. asking **your own company** to contribute code or [sponsorship of fenollp](https://github.com/sponsors/fenollp) for this upstream library [kin-openapi](https://github.com/getkin/kin-openapi) to support OpenAPI 3.1 - see here https://github.com/getkin/kin-openapi/issues/230  and specifically [this comment](https://github.com/getkin/kin-openapi/issues/230#issuecomment-1002659224).
> 
> 2. Using your company's vendor support system, report a support issue with **your company's vendor** that uses this library asking that vendor to contribute code or [sponsorship of fenollp](https://github.com/sponsors/fenollp) for this upstream library [kin-openapi](https://github.com/getkin/kin-openapi) to support OpenAPI 3.1 - see here https://github.com/getkin/kin-openapi/issues/230  and specifically [this comment](https://github.com/getkin/kin-openapi/issues/230#issuecomment-1002659224). In no particular order: Grafana, Google (Apigee), Hashicorp, PingCap, Ory, Solo, Replicate, and all the blockchain companies that own the downstream.
> 
> 3. Report a GitHub issue with [one of the many company-owned popular libraries and applications that depends on this one](https://www.grank.dev/pkg/github.com/getkin/kin-openapi.html) asking **their corporate sponsor** to contribute code or [sponsorship of fenollp](https://github.com/sponsors/fenollp) for this upstream library [kin-openapi](https://github.com/getkin/kin-openapi) to support OpenAPI 3.1 - see here https://github.com/getkin/kin-openapi/issues/230  and specifically [this comment](https://github.com/getkin/kin-openapi/issues/230#issuecomment-1002659224). Do **_not_** pester other individual unpaid volunteer maintainers of other open-source projects.
> 
> > I have been maintaining this project for a few years now and am looking to turn this into a more sustainable initiative.  Recently, I set up GitHub Sponsorship so if you'd like to support this project's future and preserve some of my mental health I encourage you to register for a monthly sponsoring [here](https://github.com/sponsors/fenollp). A one-time donation can also help solve piecemeal issues.
> > 
> > Here's a tutorial from GitHub on [how to find other maintainers that may need your help](https://docs.github.com/en/sponsors/getting-started-with-github-sponsors/quickstart-for-finding-contributors-to-sponsor) whose work you might depend on.
> > 
> > Regarding this specific issue, considering the large ramifications and attention to quality it entails, I would like to work on it once some funding has been promised or secured. We can discuss terms privately.
> > 
> > Licensing won't change due to sponsorship. All open-source work will continue to be available to all and enjoyed equally by all community members.
> 
> Just always remember:
> 
> ![dependency](https://github.com/user-attachments/assets/b215574f-6873-4ddc-b1a5-4ce0ee4989aa)
> 
> 

Thank you for notice. Meanwhile, we switched from using this library to Huma Framework.


I would have warmly supported your initiative one year ago, two years ago, but not today.

You comment a little bit late

**Comment by StevenACoffman** [comment]
@excavador :
> Thank you for notice. Meanwhile, we switched from using this library to Huma Framework.

Everyone's glad you found a solution that works for you. I hope you and your company learned to sponsor the developers of  your new choice of open-source software and its dependencies to keep them maintained.

**Comment by praneetloke** [comment]
@excavador "You comment a little bit late" is unfair and a bit short-sighted in my opinion.

> I would have warmly supported your initiative one year ago, two years ago, but not today.

@fenollp edited the issue description back in Jan 2022, more than a year before your first [comment](https://github.com/getkin/kin-openapi/issues/230#issuecomment-1866124823) (in Dec 2023) stating "Any updates?" when it was clear that support from the community was requested. It's not just you though and am not accusing any individual specifically of ignoring the support request. I understand not everyone has the means to support financially as an individual but if your company is benefiting from an open-source project, you should strongly advocate for supporting it financially.

> Thank you for notice. Meanwhile, we switched from using this library to Huma Framework.

FYI, perhaps you know this already but this issue was opened by @danielgtaylor who created Huma framework. Huma also has a [sponsors](https://github.com/danielgtaylor/huma/?tab=readme-ov-file#sponsors) section.

In the end, replies such as "I switched to another library because this maintainer is xyz to implement abc..." are not helpful at all. Feel free to switch and add the alternate library to https://github.com/getkin/kin-openapi?tab=readme-ov-file#alternatives if you wish.

**Comment by excavador** [comment]
@StevenACoffman @praneetloke 

Hello,

It seems like I did not explain properly what I meant, and you have an incorrect impression or did not understand what point I am trying to make.

It's all about expectation management and proper community communication. 
Every single personal contributor or engineer has limited resources - and limited time.
Even in the case of a company, the company also has limited resources and money.

This issue has stayed in the state "common soon" for the last two years. 
It's impossible to make any decision of planning in the current state of the issue.
People are waiting, asking "when?", because they rely on this library.

In our case, in December 2023, I considered the cost of upgrading this library to 3.1 by myself or my company's engineers.
In general, I declined this idea because of the following:
- I could observe long-standing issues and Pull Requests without reaction for weeks
- I could see, that this is not a simple task which requires a lot of changes
Meanwhile, "Huma" was in a state where it was too risky to choose Huma.

During the Summer of 2024, I reconsidered the state again. We need OpenAPI 3.1 already, and Huma, contributor and maintainer, make amazing communication with community
- fast Pull Request Acceptance
- proper communication with issues
- expectation management.

This issue of kin-openapi was in the state "coming soon," with many questions, such as "any updates?" without any answer.
It was the impression that the project was dying, first of all - because of a lack of communication.

As a result, we spent around $ 30,000 euros to migrate to Huma; meanwhile, Huma is not perfect. Also, my engineers created around 15 issues and 10 pull requests with fixes for Huma, and the authors of Huma carefully worked with their own community.

In these situations, I have zero doubts about contributing my personal time or company time and money to Huma because the maintainer of Huma is responsible.

Even the current comment—"project needs support"—is okay, guys. But what about the plan? What about the estimation of the work? What about the timeline and roadmap?

This part of communication to the community is heavily broken by kin-open API maintainers, and BECAUSE OF THAT, I prefer to invest time and money of myself and my company in Huma instead of kin-open API.

It is NOT about "support-less" from my side or company side.
It is NOT about "we are using it for free without any help from out side"

It is all about proper COMMUNICATION and EXPECTATION MANAGEMENT from both side.
Request like this one year ago will led me to question "could you please explain, how much money or work you have and how we could help us?"
But the resources (money & engineer time) which I am as manager of company could invest to kin-openapi was already invested to other things
- migration of product to Huma
- contribution to Huma from our company side

I am not blaming anybody! I just explain the ratio of decision making, and why exactly I did not see kin-openapi as the good investment.
It is not about "I am greedy", it is about "it will be bad investment" - because of the broken communication to community.

With best regards, Oleg

**Comment by StevenACoffman** [comment]
@excavador Thanks for explaining your reasoning, as your prior comment came across as unhelpful. I'm glad you and your company are contributing back to Huma and are able to collaborate with the community to the benefit of all.

Be kind to volunteer maintainers and not feel entitled to their free future work. My lesson from the past few years was to be proactive as a maintainer in asking for community help early, and to try to contribute back generously on whatever I depend on that I'd like to see continue.

**Comment by excavador** [comment]
> @excavador Thanks for explaining your reasoning, as your prior comment came across as unhelpful. I'm glad you and your company are contributing back to Huma and are able to collaborate with the community to the benefit of all.
> 
> Be kind to volunteer maintainers and not feel entitled to their free future work. My lesson from the past few years was to be proactive as a maintainer in asking for community help early, and to try to contribute back generously on whatever I depend on that I'd like to see continue.

Glad, that we clear the air and understand each other!

**Comment by daveshanley** [comment]
> Hi guys, please keep it up with supporting this library. I tried to switch to libopenapi (promised "enterprise grade" and 3.1 support), but I am back after facing some serious bugs and design problems.

Can I help? What bugs? What design problems?

**Comment by praneetloke** [comment]
@jamietanna I am curious if your [sponsorship conversation](https://github.com/getkin/kin-openapi/issues/230#issuecomment-3209288269) ever took place? I am a GH sponsor for this repo but I am an indie engineer. It would be nice if someone from a larger org could sponsor this lib. A lot of tools depend on this library. It would be great if a sponsorship could be secured to fund @fenollp's time. It's been 3 months since any PR was merged in this repo. I haven't seen his engagement in any of the issues since then either.

@fenollp we would appreciate at least a response from you if you are still looking for a sponsor and if you would be able to take a look at #1102. If you don't wish to support this project anymore, it's ok to say so. It's not fair for anyone to expect you to pour your free time forever but it would nice to know if you've moved on to other things and have no more time to dedicate for this project.

**Comment by fenollp** [comment]
Hi there!

I am still around and intend to continue maintaining this lib. I have had little time for a long while but no longer. Know that this project and this specific issue is moving forward.
Please do consider sponsoring your upstream maintainers!

Thanks everyone for your patience and kind words.

**Comment by fenollp** [comment]
I just cut https://github.com/getkin/kin-openapi/releases/tag/v0.136.0 
Please take a look and give feedback.

**Comment by jamietanna** [comment]
That's awesome news @fenollp - thanks for your work on helping get this going!

For such a significant piece of core infrastructure for Go, have you received any sponsorship for the work to support 3.1?

**Comment by fenollp** [comment]
Hey thanks for the kind words!
@reuvenharrison has been very helpful getting this out the door (thanks again!)

I did not receive any funding for 3.1 work.

**Comment by praneetloke** [comment]
Thank you, @fenollp and @reuvenharrison. For a while without activity, I thought this project was dead. I am so glad to see the renewed activity.

A BIG thank you to @reuvenharrison, as well as those that tried to get PRs for 3.1 support going prior to his mega PR.

@fenollp I suggest looking at https://floss.fund/. A number of projects have secured funding through that. https://futo.tech/grants is another one. I realize applying for grants takes a quite a bit of effort, but I think it'll be worth it in the long run.

**Comment by fenollp** [comment]
It's now been 10 releases (meaning: some time) since the release of the initial support for 3.1 and fewer and fewer 3.1 issues/PRs have been coming even after oapi-codegen adoption. Marking this as done :)
